### ISSUE TITLE
Please consider adding more optimization support for multi-layer recurrent neural networks.

### ISSUE THREAD

**Issue by wencan** [pytorch/pytorch#120072]
### 🚀 The feature, motivation and pitch

I attempted to construct a small-scale natural language processing model using torch.nn.GRU(num_layers=N, bidirectional=True). During training, the loss consistently decreased, but at a slow rate, especially after several dozen epochs, the decrease in loss was not very significant.

I changed my strategy and switched to using multiple single-layer GRUs, with LayerNorm and residual connections added in between. The improvement in fitting speed was evident.

I believe that adding more optimization support for multi-layer recurrent neural networks is necessary. Otherwise, support for multi-layer bidirectional becomes more of a hassle than a convenience.

**Comment by mikaylagawarecki** [comment]
Hey @wencan, could you elaborate on what you mean by more optimization support for multi-layer recurrent neural networks please?

Perhaps a code sample of what you had to do before and after would also be helpful

**Comment by wencan** [comment]
@mikaylagawarecki 
A network structure with slow fitting speed:
```
torch.nn.GRU(input_size=*,
                         hidden_size=*,
                         num_layers=2,  # ⚠️two layers
                         batch_first=True,
                         dropout=*,
                         bidirectional=True
)
```

improved network structure:
```
        pool_out = torch.clone(out)

        self.gru_1 = torch.nn.GRU(input_size=self._gru_1_input_size,
                                hidden_size=self._gru_1_hidden_size//2,
                                num_layers=1,  # ⚠️one layer
                                batch_first=True,
                                bidirectional=True)
        self.res_block_2 = GatedResidualBlock('avg', self._gru_1_hidden_size) # inputs: pool outputs, gru_1 outputs
        self.layer_norm_2 = torch.nn.LayerNorm(self._gru_1_hidden_size)
        self.dropout_2 = torch.nn.Dropout(self._dropout_2)

        self.gru_2 = torch.nn.GRU(input_size=self._gru_2_input_size,
                                hidden_size=self._gru_2_hidden_size//2,
                                num_layers=1,  # ⚠️one layer
                                batch_first=True,
                                bidirectional=True)
        self.res_block_3 = GatedResidualBlock('avg', self._gru_2_hidden_size)    # inputs: gru_1 outputs, gru_2 outputs
        self.layer_norm_3 = torch.nn.LayerNorm(self._gru_2_hidden_size)
        self.dropout_3 = torch.nn.Dropout(self._dropout_3)

# ... ... 

        out = torch.nn.utils.rnn.pack_padded_sequence(out, lengths, batch_first=True, enforce_sorted=False)
        out, _ = self.gru_1(out)
        out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # -> (batch_size, max_length, gru_hidden_size)
        out = self.res_block_2(pool_out, out)
        out = self.layer_norm_2(out)
        out = torch.nn.functional.gelu(out)
        out = self.dropout_2(out)
        gru_1_out = torch.clone(out)

        out = torch.nn.utils.rnn.pack_padded_sequence(out, lengths, batch_first=True, enforce_sorted=False)
        out, _ = self.gru_2(out)
        out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # -> (batch_size, max_length, gru_hidden_size)
        out = self.res_block_3(gru_1_out, out)
        out = self.layer_norm_3(out)
        out = torch.nn.functional.gelu(out)
        out = self.dropout_3(out)


```
